Cabal/Distribution/Simple/Program/Find.hs

```diff
@@ -109,7 +109,7 @@ findProgramOnSearchPath verbosity searchpath prog = do
     findFirstExe (f:fs) = do
       isExe <- doesExecutableExist f
       if isExe
-        then return (Just f)
+        then Just `fmap` canonicalizePath f
```

Comment:
It only returns absolute paths if all of the `ProgramSearchPathDir`s are absolute. That's the whole point. You'll notice my change is the only call to `canonicalizePath` in the function.